### PR TITLE
Extension of `OPENSSL_ia32cap` to accommodate additional `CPUID` feature/capability bits

### PR DESCRIPTION
**Pull request by aelizaro**
Extended `OPENSSL_ia32cap` support to accommodate additional `CPUID`:
 bits 128 - 191 CPUID.(EAX=07H,ECX=0H).EDX and CPUID.(EAX=07H,ECX=1H).EAX
 bits 192 - 255 CPUID.(EAX=07H,ECX=1H).EDX and CPUID.(EAX=07H,ECX=1H).EBX
 bits 256 - 319 CPUID.(EAX=07H,ECX=1H).ECX and CPUID.(EAX=24H,ECX=0H).EBX
 
 Output for Intel SDE:
```bash
# before:
$ sde -spr -- openssl version -a
CPUINFO: OPENSSL_ia32cap=0x7ffef3ffffebffff:0xfb417ffef3bfb7ef

# after:
$ sde -spr -- openssl version -a
CPUINFO: OPENSSL_ia32cap=0x7ffef3ffffebffff:0xfb417ffef3bfb7ef:0x00001c30ffdc4432:0x0000000000000000:0x0000000000000000
```

E.g. Lunar Lake is a hybrid CPU (bit 128 + 15) supports SHA512, SM3 and SM4 instructions, which are bits 0, 1, 2 of `CPUID.(EAX=07H,ECX=1H).EAX`, which corresponds to bits 128+32...128+34 and AVX-IFMA bit 128 + 55 (highlighted below)

```
$ sde -lnl -- openssl version -a
OPENSSL_ia32cap=0x7ffaf3ffffebffff:0x9bc007bc239cb7ab:0x00c0009700145030:0x0000000200000430:0x0000000000000000
```

> bits 191 - 128 (right to left): 0x00**c**0009**7**0014***5***030 = 0x0000'0000'1***1***00'0000'0000'0000'1001'0**111**'0000'0000'0001'0100'0***1***01'0000'0011'0000

Granite Rapids has AVX10 support (256+32) with XMM (256+48), YMM (256+49) and ZMM (256+50) registers

```
$ sde -gnr -- openssl version -a
OPENSSL_ia32cap=0x7ffaf3ffffebffff:0xbb405ffef3bfb7af:0x0060003003d45030:0x0000000000084100:0x0007000100000000
```

> bits 319 - 256 0x0007000100000000 = 0x0000'0000'0000'0***111***'0000'0000'0000'000***1***'0000'0000'0000'0000'0000'0000'0000'0000

##### Checklist
- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/man3/OPENSSL_ia32cap.pod
Comment (stanciuadrian) [inline doc/man3/OPENSSL_ia32cap.pod:31]: Running `make doc-nits` gives 6 warnings, all related to `=head3`:

```
*** WARNING: =head3 without preceding higher level at line 31 in file doc/man3/OPENSSL_ia32cap.pod
*** WARNING: =head3 without preceding higher level at line 79 in file doc/man3/OPENSSL_ia32cap.pod
*** WARNING: =head3 without preceding higher level at line 116 in file doc/man3/OPENSSL_ia32cap.pod
*** WARNING: =head3 without preceding higher level at line 138 in file doc/man3/OPENSSL_ia32cap.pod
*** WARNING: =head3 without preceding higher level at line 152 in file doc/man3/OPENSSL_ia32cap.pod
*** WARNING: =head3 without preceding higher level at line 170 in file doc/man3/OPENSSL_ia32cap.pod
```

It looks like `=head2` is missing and some (or all) level 3 sections may need to be upgraded to level 2.

File: crypto/cpuid.c
Comment (stanciuadrian) [inline crypto/cpuid.c:130]: An explicit `NULL` test for `env` (similar to the one 2 lines below) can increase the readability.

File: crypto/cpuid.c
Comment (stanciuadrian) [inline crypto/cpuid.c:128]: Declaring loop variables before the `for` statement is a good way to maintain C90 compatibility.

**Comment by stanciuadrian** [comment]
General `git` workflow comment: squashing/rebasing often should clean the PR and make the changes easier to understand.

File: crypto/cpuid.c
Comment (tkanteck) [inline crypto/cpuid.c:143]: good idea to create a loop across the indexes to address comment about code duplication

File: doc/man3/OPENSSL_ia32cap.pod
Comment (stanciuadrian) [inline doc/man3/OPENSSL_ia32cap.pod:211]: Perhaps it's a good idea to extend this kind of patch to all the files from this PR.

Comment (aelizaro) [reply doc/man3/OPENSSL_ia32cap.pod]: Thanks, removed this change

File: crypto/cpuid.c
Comment (aelizaro) [reply crypto/cpuid.c]: Added a check

File: doc/man3/OPENSSL_ia32cap.pod
Comment (aelizaro) [reply doc/man3/OPENSSL_ia32cap.pod]: Thank you for catching this, fixed!

File: CHANGES.md
Comment (stanciuadrian) [inline CHANGES.md:75]: This well-written summary can be further shortened to replace the PR title and/or the first line of the commit message.

Comment (aelizaro) [reply CHANGES.md:75]: Thank you, updated the commit message

File: NOTES-VALGRIND.md
Comment (stanciuadrian) [inline NOTES-VALGRIND.md:54]: It looks like the original `OPENSSL_ia32cap=":0"` doesn't disable post AVX2 instructions anymore.
This could behave as a breaking change for some setups.

**Review by stanciuadrian — APPROVED** [review]
The PR is in a very good shape! Thank you Alina and Dan for preparing this PR.
Will help a lot with maintainability and future proofing.

The documentation is comprehensive and easy to understand/use.

File: NOTES-VALGRIND.md
Comment (aelizaro) [reply NOTES-VALGRIND.md:54]: It is easy to adjust with a loop here: https://github.com/openssl/openssl/blob/c87ea2e7b6266e2fc2d7b2aecefa276c96a767cf/crypto/cpuid.c#L145 to zeroize the rest of indexes if needed.

If the adjustment of `OPENSSL_ia32cap` environment variable is a very specific action and it suppose to be manual and explicit - I would advise to keep it as, so the users know and understand what they are turning on and off. There is some past experience from the customers that the implicit disable ‘feature’ had teams doing a lot of debugging, to figure out the RSA performance anomaly.  Ultimately it was just the ‘implicit’ zeroing of `OPENSSL_ia32cap` vectors that caused the problem.

On the other hand, if it is common and advisable to use `OPENSSL_ia32cap=":0"` for the debug, maybe it is worth to keep the old behavior as and make it better documented. However, it may lead to some unexpected results on any post AVX2 HW.

@t8m, do you have any thoughts on how it is better to proceed from OpenSSL maintainers experience?

**Comment by aelizaro** [comment]
@t8m, I assume the PR is ready for the review from the OpenSSL maintainers as we got two approvals from @tkanteck  and @stanciuadrian (thank you very much for the review!)

File: NOTES-VALGRIND.md
Comment (t8m) [reply NOTES-VALGRIND.md:54]: IMO we need to keep the backwards compatibility for this.

Comment (aelizaro) [reply NOTES-VALGRIND.md]: @t8m, I've updated the behavior for the environment variable and documentation, please, check if that is what expected: https://github.com/openssl/openssl/pull/25709/commits/0c81445cc4b108434c18a7c1ffadecaccca3895d
Now it should not break the expected behavior for cases as OPENSSL_ia32cap=":0" - all the following indexes will be set to 0. It also works for example like:
OPENSSL_ia32cap="::0" - zeroize starting 3rd value or OPENSSL_ia32cap=":0::::" - zeroize only the second one and keep the remaining untouched.

My only concern now is for the situation if OPENSSL_ia32cap would be expanded again, then OPENSSL_ia32cap=":0::::" will zeroize the new indexes after existing once, however, this looks like a really exotic case to turn off all AVX2 and leave the rest unchanged.

File: crypto/cpuid.c
Comment (mattcaswell) [reply crypto/cpuid.c:134]: All variables must be declared at the start of a block - no mixed declarations and code.

File: crypto/cpuid.c
Comment (mattcaswell) [inline crypto/cpuid.c:130]: The initial `env != NULL` here makes no sense. We have already derefed `env` in the code above, and confirmed that `env` is non NULL in the `ossl_get_env` call.

File: crypto/cpuid.c
Comment (mattcaswell) [inline crypto/cpuid.c:131]: Isn't this always true if we get here?

File: crypto/x86cpuid.pl
Comment (mattcaswell) [inline crypto/x86cpuid.pl:151]:
```suggestion
	&mov	(&DWP(32,"edi"),"ecx");	# save cpuid(EAX=0x7, ECX=0x1).ECX to OPENSSL_ia32cap_P[8]
```

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/cpuid.c
Comment (aelizaro) [reply crypto/cpuid.c]: yes, thank you for catching this, removed

File: crypto/cpuid.c
Comment (aelizaro) [reply crypto/cpuid.c]: removed extra check

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping @openssl/committers for second review

**Comment by mattcaswell** [comment]
Hmmmm....the CI test failure on Windows is strange....

**Comment by t8m** [comment]
> Hmmmm....the CI test failure on Windows is strange....

That's a known and unrelated issue. Unfortunately I could not find time yet to properly investigate it.

**Comment by aelizaro** [comment]
> > Hmmmm....the CI test failure on Windows is strange....
> 
> That's a known and unrelated issue. Unfortunately I could not find time yet to properly investigate it.

Please, let me know when to rebase again to be 100% sure CI is green (at least it was green on the previous rebase)

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**
